std/typecons.d

```diff
@@ -5893,3 +5893,143 @@ public:
     auto value = cast(int)flags_A;
     assert(value == Enum.A);
 }
+
+/**
+   Type constructor for head-const variables.
+
+   Head-const variables cannot be directly mutated or rebound, but references
+   reached through the variable are typed with their original mutability.
+
+   It is equivalent to $(D final) variables in D1 and Java,
+   as well as $(D readonly) variables in C#.
+ */
+struct HeadConst(T)
+    if(!is(T == const) && !is(T == immutable))
+{
+    private:
+    T data;
+
+    public:
+    /// $(D HeadConst) subtypes $(D T) as an rvalue.
+    inout(T) HeadConst_get() inout pure nothrow @nogc @safe
+    {
+        return data;
+    }
+
+    alias HeadConst_get this; /// Ditto
+
+    ///
```

Comment:
I think they're fairly self-documenting, but I'll think of something useful to add.